### PR TITLE
Product Collection: Make it compatible with filter blocks

### LINKED ISSUE #9876: Product Collection - should work well with Filter blocks
### Description

At the moment Filters don't work with Product Collection. They don't appear on the page even if Product Collection inherits from a global query which is a similar behavior to when there's no products on the page

### Goal

Make sure Product Collection is working well with the Filter blocks.

### PR DESCRIPTION
**Pull request by imanish003**
This PR introduces following changes:
1. All filter blocks now work with Product Collection block i.e. price filter, attributes filter, stock status filter, and rating filter blocks.
2. **Renaming and extending existing methods**: Some methods are renamed for better clarity, for example `build_query` to `build_frontend_query` and `update_rest_query` to `update_rest_query_in_editor`. 

Fixes #9876 

#### Test that filter blocks work with the Product Collection block

1. Create a new post and add a 'Product Collection' block to it. 
2. Add all the filter blocks to the post. Alternatively, you can use the 'Product Filters' pattern to add all filter blocks swiftly. 
3. Save the post and navigate to the frontend. 
4. Confirm that all filters function in harmony with the 'Product Collection' block.

Here is a quick demo:

https://github.com/woocommerce/woocommerce-blocks/assets/16707866/e4215e0f-d7bb-4039-8cff-4187ef809096

#### Test with Product Catalog template

1. Open the product catalog template within the editor.
2. Insert a 'Product Collection' block into the Product Catalog. When the 'Product Collection' block is added to the Product Catalog template, the 'Inherit query from template' option will be enabled by default. Keep it enabled. 
3. Save the changes and navigate to '/shop'.
4. Confirm that all product filter blocks operate as expected.

Here is a quick demo:

https://github.com/woocommerce/woocommerce-blocks/assets/16707866/6caa9b83-6093-471e-82a7-f14ed1416bbd

* [X] Do not include in the Testing Notes 

### WooCommerce Visibility

* [ ] WooCommerce Core
* [ ] Feature plugin
* [X] Experimental

### Changelog

> Product Collection: Make it compatible with filter blocks

### CODE REVIEW

File: assets/js/utils/filters.ts
Comment (imanish003) [inline assets/js/utils/filters.ts:1]: FYI It was discussed on slack p1687253475483319-slack-C02FL3X7KR6 that if any change is made to filters, then we should bring back user to 1st page.

**Comment by imanish003** [comment]
Hi @Aljullu, could you please review the changes made to the `assets/js/utils/filters.ts` file in this PR? These modifications also have an impact on the following blocks:

- 'Classic WooCommerce Product Grid Block'
- 'Products Beta' block

Your feedback on these changes would be greatly appreciated. Thank you! 🙇‍♂️

**Review by kmanijak — APPROVED** [review]
Great work! I've tested with Product Collection, but also the page change with Products and Classic Template block, and all works well.

I would only change "Do not include in the Testing Notes" or add a not that the last part (about defaulting to the first page for Products block) should be included in the testing notes of the release. What do you think?

**Comment by Aljullu** [comment]
> Hi @Aljullu, could you please review the changes made to the `assets/js/utils/filters.ts` file in this PR?

Sure thing! I see these changes have been removed from this PR. Is the plan to add them to another PR?

**Comment by imanish003** [comment]
@Aljullu I apologize for any confusion caused, Actually I discussed with @kmanijak and moved those changes to new PR: https://github.com/woocommerce/woocommerce-blocks/pull/9952 🙂

This was done to separate "WooCommerce Core" & "Experimental" changes. Also, It will be helpful for the release manager too.

**Pull request merged**